### ISSUE TITLE
Enhancement: Learning rate, plateauing while training, implement early stopping

### ISSUE THREAD

**Issue by breznak** [dbolya/yolact#298]
When training on a custom dataset, we typically reach a point when YOLACT achieves certain accuracy, and training does not improve further till the end of training (N steps)
- `train.py` has a rather simple `learning rate` adjustment strategy (different alpha at defined 4 steps)
- good thing is the net does `not overfit` (train & eval accuracy does not diverge, both not worsening)
- this issue is not about achieving better results, but avoiding wasting time when we don't already improve further. 

I'd like to propose `early-stopping` as known eg. from tensorflow models. Definy stopping strategy as: 
- reached max num steps (default 800000 in train.py)
- reached desired accuracy (if we're good enough with certain acc, no point trainning further)
- plateauing: ("plateau is when accuracy does not improve for a while)
  - plateau detected if accuracy is not changed over (P) epochs
  - optional: if so, we can do some magic with learning-rate to avoid it
  - else: stop early.

**Comment by dbolya** [comment]
Yup, I wanted to implement this too but then I went on break. There's also a paper that has a formula for an optimal learning rate for super quick convergence that I want to give a try.

I'll add this to a high spot on my TODO list.

**Comment by breznak** [comment]
Here are some sources on LR & training management: 
https://github.com/pytorch/ignite/issues/560#issuecomment-513114197 
Would it be OK to include `pytorch-ignite`? Looks like what I had on my mind: 
https://pytorch.org/ignite/quickstart.html

**Comment by dbolya** [comment]
I took a look over it, and it has useful features but probably not for this project. I'm not sure I want to fit my existing code into a completely new framework just to add early stopping. Couldn't we just add it as a standalone function (maybe as a custom pytorch optimizer)?

**Comment by breznak** [comment]
> not sure I want to fit my existing code into a completely new framework just to add early stopping. 

I think this standartization would be ideal long-term but agree it's a lot of work. And if talking about early-stop, yep, you're right. 

> Couldn't we just add it as a standalone function (maybe as a custom pytorch optimizer)?

:+1: This would be a good compromise. Does early-stop (possibly later other things with LR ("preheating", ...)), not too much rework (1 new class), future-proof (pytorch standard)
